core/network/src/consensus_gossip.rs

```diff
@@ -293,65 +339,63 @@ impl<B: BlockT> ConsensusGossip<B> {
-			let engine_id = message.engine_id;
-			// validate the message
-			let (topic, status) = match self.validators.get(&engine_id)
-				.map(|v| v.validate(&message.data))
-			{
-				Some(ValidationResult::Valid(topic)) => (topic, Status::Live),
-				Some(ValidationResult::Future(topic)) => (topic, Status::Future),
-				Some(ValidationResult::Invalid) => {
-					trace!(target:"gossip", "Invalid message from {}", who);
-					protocol.report_peer(
-						who,
-						Severity::Bad(format!("Sent invalid consensus message")),
-					);
-					return None;
-				},
-				Some(ValidationResult::Expired) => {
-					trace!(target:"gossip", "Ignored expired message from {}", who);
-					return None;
-				},
-				None => {
-					protocol.report_peer(
-						who.clone(),
-						Severity::Useless(format!("Sent unknown consensus engine id")),
-					);
-					trace!(target:"gossip", "Unknown message engine id {:?} from {}",
-						engine_id, who);
-					return None;
-				},
-			};
+		let engine_id = message.engine_id;
+		//validate the message
+		let validation = self.validators.get(&engine_id)
+			.cloned()
+			.map(|v| {
+				let mut context = ValidatorContext { gossip: self, protocol, engine_id };
+				v.validate(&mut context, &who, &message.data)
+			});
+		let validation_result = match validation {
+			Some(ValidationResult::KeepAndPropagate(topic)) => Some((topic, true)),
+			Some(ValidationResult::ProcessAndDiscard(topic)) => Some((topic, false)),
+			Some(ValidationResult::Discard) => None,
+			None => {
+				trace!(target:"gossip", "Unknown message engine id {:?} from {}", engine_id, who);
+				protocol.report_peer(
+					who,
+					Severity::Useless(format!("Sent unknown consensus engine id")),
+				);
+				return;
+			}
+		};
 
-			peer.known_messages.insert(message_hash);
-			if let Entry::Occupied(mut entry) = self.live_message_sinks.entry((engine_id, topic)) {
-				debug!(target: "gossip", "Pushing consensus message to sinks for {}.", topic);
-				entry.get_mut().retain(|sink| {
-					if let Err(e) = sink.unbounded_send(message.data.clone()) {
-						trace!(target:"gossip", "Error broadcasting message notification: {:?}", e);
+		if let Some((topic, keep)) = validation_result {
+			if let Some(ref mut peer) = self.peers.get_mut(&who) {
+				use std::collections::hash_map::Entry;
+				peer.known_messages.insert(message_hash);
+				if let Entry::Occupied(mut entry) = self.live_message_sinks.entry((engine_id, topic)) {
+					debug!(target: "gossip", "Pushing consensus message to sinks for {}.", topic);
+					entry.get_mut().retain(|sink| {
+						if let Err(e) = sink.unbounded_send(TopicNotification {
+							message: message.data.clone(),
+							sender: Some(who.clone())
+						}) {
+							trace!(target: "gossip", "Error broadcasting message notification: {:?}", e);
+						}
+						!sink.is_closed()
+					});
+					if entry.get().is_empty() {
+						entry.remove_entry();
+					}
+					if keep {
+						self.register_message(message_hash, topic, message);
```

Comment:
`register_message` doesn't propagate the message. There's no propagation happening in this method.

Reply:
Yes, that's intentional -- the gossip validator is meant to trigger propagation when appropriate.

Reply:
We should maybe rename the `ValidationResult` enum then. `ProcessAndKeep`, `ProcessAndDiscard`, `Discard` ?